### PR TITLE
Miscellaneous low-level performance improvements

### PR DESCRIPTION
**Pull request by SamGondelman**
Based on hotspots detected with Visual Studio's debugger and VTune.

Test plan:
- Verify that blendshapes still work.  Talk to someone else.  Both of your mouths/eyes/faces should move correctly.
- Verify that animated models still animate (e.g. flags in dev-welcome).
- In engine-dev, verify that closer avatars animate slightly faster than those that are farther away.
- Before avatar models load, the pink sphere should render as before.  Once the avatar appears, the pink sphere should disappear.
- On Android, go to a domain with a few other people (not several hundred).  Verify that blendshapes on other people work.

### CODE REVIEW

File: interface/src/avatar/OtherAvatar.cpp [hunk 1]
```diff
@@ -29,20 +29,23 @@ OtherAvatar::~OtherAvatar() {
 }
 
 void OtherAvatar::removeOrb() {
-    if (qApp->getOverlays().isAddedOverlay(_otherAvatarOrbMeshPlaceholderID)) {
```
Comment (SamGondelman) [inline interface/src/avatar/OtherAvatar.cpp]: isAddedOverlay performs a blocking invoke, a mutex lock, and two map lookups.

File: libraries/shared/src/AABox.cpp [hunk 2]
```diff
@@ -79,33 +79,23 @@ void AABox::setBox(const glm::vec3& corner, const glm::vec3& scale) {
...
-    if (normal.z < 0.0f) {
-        result.z += _scale.z;
-    }
-
+    float blend = (float)(normal.x < 0.0f);
+    result.x += blend * _scale.x + (1.0f - blend) * 0.0f;
+    blend = (float)(normal.y < 0.0f);
+    result.y += blend * _scale.y + (1.0f - blend) * 0.0f;
+    blend = (float)(normal.z < 0.0f);
+    result.z += blend * _scale.z + (1.0f - blend) * 0.0f;
```
Comment (SamGondelman) [inline libraries/shared/src/AABox.cpp]: branchless

File: libraries/shared/src/AABox.h [hunk 3]
```diff
@@ -85,8 +85,21 @@ class AABox {
...
     AABox clamp(float min, float max) const;
 
-    AABox& operator += (const glm::vec3& point);
-    AABox& operator += (const AABox& box);
+    inline AABox& operator+=(const glm::vec3& point) {
+        float blend = (float)isInvalid();
+        glm::vec3 maximumScale(glm::max(_scale, point - _corner));
+        _corner = glm::min(_corner, point);
+        _scale = blend * _scale + (1.0f - blend) * maximumScale;
+        return (*this);
```
Comment (SamGondelman) [inline libraries/shared/src/AABox.h]: This was a hotspot before.  Now it's: inline, branchless, and simpler.

File: libraries/shared/src/GLMHelpers.h [hunk 4]
```diff
@@ -316,4 +316,12 @@ inline void glm_mat4u_mul(const glm::mat4& m1, const glm::mat4& m2, glm::mat4& r
 #endif
 }
 
+inline glm::vec3 fastRoundf(const glm::vec3& vec) {
+#if GLM_ARCH & GLM_ARCH_SSE2_BIT
+    return glm::vec3(_mm_cvt_ss2si(_mm_set_ss(vec.x)), _mm_cvt_ss2si(_mm_set_ss(vec.y)), _mm_cvt_ss2si(_mm_set_ss(vec.z)));
```
Comment (SamGondelman) [inline libraries/shared/src/GLMHelpers.h]: 30 times faster!

File: libraries/shared/src/PrioritySortUtil.h [hunk 5]
```diff
@@ -153,7 +152,7 @@ namespace PrioritySortUtil {
         }
 
         ConicalViewFrustums _views;
-        std::priority_queue<T> _queue;
+        std::vector<T> _vector;
```
Comment (SamGondelman) [inline libraries/shared/src/PrioritySortUtil.h]: Sorting a vector after adding everything instead of inserting into a priority queue and then repeatedly popping the top is ~2.5 times faster (5 times faster on Andrew's)

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12541/HighFidelity-Beta-PR13914-5f08ed5.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12541/HighFidelity-Beta-PR13914-5f08ed5.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12541/HighFidelity-Beta-Interface-PR13914-5f08ed5.dmg

**Comment by SamGondelman** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12551/HighFidelity-Beta-PR13914-5f08ed5.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12551/HighFidelity-Beta-PR13914-5f08ed5.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12551/HighFidelity-Beta-Interface-PR13914-5f08ed5.dmg

**Comment by SamGondelman** [comment]
build this please

**Comment by qlipdahlin** [comment]
Currently Testing

File: interface/src/avatar/AvatarManager.cpp [hunk 6]
```diff
@@ -260,26 +260,17 @@ void AvatarManager::updateOtherAvatars(float deltaTime) {
             // --> some avatar velocity measurements may be a little off
 
             // no time to simulate, but we take the time to count how many were tragically missed
-            bool inView = sortData.getPriority() > OUT_OF_VIEW_THRESHOLD;
-            if (!inView) {
```
Comment (ZappoMan) [inline interface/src/avatar/AvatarManager.cpp]: what is this old logic doing? it seems to be missing from the new code. It's not clear to me if we intended to remove this... if so, why?

File: interface/src/avatar/OtherAvatar.cpp (hunk 1 above)
Comment (ZappoMan) [reply interface/src/avatar/OtherAvatar.cpp]: oh jez!

File: libraries/shared/src/AABox.cpp [hunk 7]
```diff
@@ -79,33 +79,23 @@ void AABox::setBox(const glm::vec3& corner, const glm::vec3& scale) {
...
-        result.x += _scale.x;
-    }
-    if (normal.y > 0.0f) {
-        result.y += _scale.y;
-    }
-    if (normal.z > 0.0f) {
-        result.z += _scale.z;
-    }
+    float blend = (float)(normal.x > 0.0f);
+    result.x += blend * _scale.x + (1.0f - blend) * 0.0f;
```
Comment (ZappoMan) [inline libraries/shared/src/AABox.cpp]: wow? really... I'm definitely surprised this is faster, but I'm sure Ken would have written it this way to begin with. 

Also, I haven't had any coffee yet this morning, but how is `(1.0f - blend) * 0.0f` ever gonna be anything except `0.0f`?

For us mere mortals... can you please add a comment that "explains" the old logical form of this... so people don't have to parse the math to realize this is only adding the scale if the normal is greater than zero... thanks.

File: libraries/shared/src/AABox.cpp (hunk 2 above)
Comment (ZappoMan) [reply libraries/shared/src/AABox.cpp]: please add a brief comment to this code explaining the original logic/intended behavior so that mere mortals will understand your genius.

File: libraries/shared/src/AABox.h (hunk 3 above)
Comment (ZappoMan) [reply libraries/shared/src/AABox.h]: cool

**Comment by qlipdahlin** [comment]
@conklin94122 PASS

- Avatar Eyes/Mouth/Head worked correctly.
- Models are properly animating still (flags for example)
- Dev-Engine was missing avatars, used dev-mobile-master instead. Verified that closer avatars render and animate more cleanly then ones further away
- Pink sphere appears before avatar finishes loading and disappears when avatar loads

File: libraries/shared/src/GLMHelpers.h (hunk 4 above)
Comment (kencooke) [inline libraries/shared/src/GLMHelpers.h]: A couple ways to make this even better:
1) It's not obvious (due to template types) but this converts the result back to vec3<float> but we actually use int.
2) lrintf() would be better fallback than roundf() because it matches exactly (rounds to nearest even) which is desirable for testing. But both functions are slow because they set errno on overflow. Since we already require IEEE-754 float, a fast scalar function like below would also be useful elsewhere:

```
// convert float to int, using round-to-nearest-even (undefined on overflow)
inline int fastLrintf(float x) {
#if GLM_ARCH & GLM_ARCH_SSE2_BIT
    return _mm_cvt_ss2si(_mm_set_ss(x));
#else
    // return lrintf(x);
    static_assert(std::numeric_limits<double>::is_iec559, "Requires IEEE-754 double precision format");
    union { double d; int64_t i; } bits = { (double)x };
    bits.d += (3ULL << 51);
    return (int)bits.i;
#endif
}
```

File: tests/shared/src/GLMHelpersTests.cpp [hunk 8]
```diff
@@ -214,3 +214,39 @@ void GLMHelpersTests::testGenerateBasisVectors() {
         QCOMPARE_WITH_ABS_ERROR(w, z, EPSILON);
     }
 }
+
+void GLMHelpersTests::roundPerf() {
+    const int NUM_VECS = 1000000;
+    const float MAX_VEC = 500.0f;
+    std::vector<glm::vec3> vecs;
+    vecs.reserve(NUM_VECS);
+    for (int i = 0; i < NUM_VECS; i++) {
```
Comment (kencooke) [inline tests/shared/src/GLMHelpersTests.cpp]: It might be overkill, but for single variable functions it's feasible to validate over all input values.
In this case, you can test all floats (that do not overflow when converted to int) with:
```
    for (int32_t bits = -0x31000000; bits <= 0x4f000000; bits += (bits < 0 ? -1 : +1)) {
        float f = *(float*)&bits;
```

File: libraries/shared/src/AABox.cpp (hunk 7 above)
Comment (kencooke) [reply libraries/shared/src/AABox.cpp]: I like that he left the (* 0.0f) to make the idiom clear.
He's using the common mix/lerp as a branchless select.

File: interface/src/avatar/AvatarManager.cpp (hunk 6 above)
Comment (SamGondelman) [reply interface/src/avatar/AvatarManager.cpp]: I removed this intentionally because I thought it wasn't doing what we wanted but I realize now that it was trying to save us the work of looping over later avatars that won't be in view.  I'll put it back.

**Comment by conklin94122** [comment]
Removing QA Ready due to dev change requests. Please relabel when ready and we will retest.

File: tests/shared/src/GLMHelpersTests.cpp (hunk 8 above)
Comment (SamGondelman) [reply tests/shared/src/GLMHelpersTests.cpp]: I'm going to leave this as is for now.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12561/HighFidelity-Beta-PR13914-d01b438.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12561/HighFidelity-Beta-PR13914-d01b438.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12561/HighFidelity-Beta-Interface-PR13914-d01b438.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12561/HighFidelity-Beta-PR13914-d01b438.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12561/HighFidelity-Beta-Interface-PR13914-d01b438.dmg

File: tests/shared/src/GLMHelpersTests.cpp (hunk 8 above)
Comment (kencooke) [reply tests/shared/src/GLMHelpersTests.cpp]: Fair 'nuff. Just beware that you likely aren't testing corner cases like exactly xxx.5

File: interface/src/avatar/AvatarManager.cpp (hunk 6 above)
Comment (AndrewMeadows) [reply interface/src/avatar/AvatarManager.cpp]: That is correct.  It is an early exit that bails when the remaining sorted data is obviously out of view.

File: libraries/shared/src/AABox.h (hunk 3 above)
Comment (AndrewMeadows) [reply libraries/shared/src/AABox.h]: :1st_place_medal:

File: libraries/shared/src/PrioritySortUtil.h (hunk 5 above)
Comment (AndrewMeadows) [reply libraries/shared/src/PrioritySortUtil.h]: Maybe we could also `reserve(size)` of the vector before pushing things onto it?  We usually know how many things we're working with (in the context that uses the `PrioritySortUtil` API), right?

**Comment by qlipdahlin** [comment]
Currently Testing

**Comment by qaAdamSmith82** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12626/HighFidelity-Beta-PR13914-ac4b0cf.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12626/HighFidelity-Beta-PR13914-ac4b0cf.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12626/HighFidelity-Beta-Interface-PR13914-ac4b0cf.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12631/HighFidelity-Beta-PR13914-ade58a5.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12631/HighFidelity-Beta-PR13914-ade58a5.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12631/HighFidelity-Beta-Interface-PR13914-ade58a5.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12631/HighFidelity-Beta-PR13914-ade58a5.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12631/HighFidelity-Beta-Interface-PR13914-ade58a5.exe

**Comment by qlipdahlin** [comment]
@conklin94122 FAIL

Noticed shading issue in serverless. Some walls are not properly rendering when within certain range of the user, thus allowing entities like the name sign or boxes to render shows where it should not be capable of rendering it. Saw on an Intel/Nvidia and an AMD/AMD system, but was not present on the same systems and other systems running a Master build from today. Fails PR.

Screenshot - 
shadowerror

- Avatar Eyes/Mouth/Head worked correctly.
- Models are properly animating still (flags for example)
- Dev-Engine was missing avatars, used dev-mobile-master instead. Verified that closer avatars render and animate more cleanly then ones further away
- Pink sphere appears before avatar finishes loading and disappears when avatar loads
- Android saw Eyes/Mouth/Head working correctly, though it did take some time before the mouth worked correctly.

There was also a problem with a Vive system constantly crashing on PR13914-d01b438, but PR13914-ade58a5 did not appear to have the same issues.

**Comment by SamGondelman** [comment]
@conklin94122 @qlipdahlin hopefully fixed now

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/12647/HighFidelity-Beta-PR13914-9e76c16.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12647/HighFidelity-Beta-PR13914-9e76c16.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/12647/HighFidelity-Beta-Interface-PR13914-9e76c16.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12647/HighFidelity-Beta-PR13914-9e76c16.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/12647/HighFidelity-Beta-Interface-PR13914-9e76c16.exe

**Comment by qlipdahlin** [comment]
Currently Testing
